tests/models/shadow/broker.p

```diff
@@ -0,0 +1,97 @@
+//
+// The broker handles requests like Produce and Fetch.
+//
+
+// produce_request - append a batch to a partition
+//
+// source: the machine sending the request
+// request_id: must be unique per source.
+// partition: target partition for the operation
+// batch_id: the batch "content".
+//
+// behavior of the system is not affected by the content of a batch. but it is
+// still useful to generate unique batch "content" for use in assertions.
+type produce_request = (
+  source: machine,
+  request_id: int,
+  partition: Partition,
+  batch_id: int);
+
+// produce_response - returns a produced batch's offset
+//
+// request_id: value from the request. used for correlation.
+// offset: offset at which the batch in the request is located.
+type produce_response = (
+  request_id: int,
+  offset: int);
+
+// fetch_request - request a batch from a partition
+//
+// source: the machine sending the request
+// request_id: correlation id. unique per source.
+// partition: partition to read from
+// offset: offset of the batch to read
+type fetch_request = (
+  source: machine,
+  request_id: int,
+  partition: Partition,
+  offset: int);
+
+// fetch_response - returns the fetched batch
+//
+// request_id: correlation id from request
+// batch_id: the batch "content"
+type fetch_response = (
+  request_id: int,
+  batch_id: int);
+
+// Produce request and response events
+event produce_request_event : produce_request;
+event produce_response_event : produce_response;
+
+// Fetch request and response events
+event fetch_request_event : fetch_request;
+event fetch_response_event : fetch_response;
+
+// Internal event used by the broker to drive the commit protocol.
+event broker_commit_event;
+
+machine Broker {
+  var storage: Storage;
+  // received produce requests that will be committed together
+  var produce_requests: seq[produce_request];
+  var request_id: int;
+
+  start state Init {
+    entry (config: (storage: Storage)) {
+      storage = config.storage;
+      goto WaitForRequest;
+    }
+  }
+
+  state WaitForRequest {
+    on produce_request_event do (request: produce_request) {
+      // produce requests are queued and processed later.
+      produce_requests += (sizeof(produce_requests), request);
+
+      // if the pending set just transitioned to the non-empty state then ensure
+      // that it is eventually processed (there may be no more produce requests
+      // to drive execution). scheduled processing for "later" captures the
+      // non-determinism of a real system either having reached a maximum
```

Comment:
Not clear what type of non-determinism you are interested in here. Please specify for future readers. In a real system we can as well guard with a mutex the point from batch being ready to scheduling the commit event.

It this to allow the broker state machine to potentially do other work (in the case we eventually extend it)?

Reply:
> It this to allow the broker state machine to potentially do other work (in the case we eventually extend it)?

I don't think so. I clarified this comment in more detail here https://github.com/redpanda-data/redpanda/pull/23053/commits/e612963e2b381045b89c00b289f4055fcfeeb517.

Reply:
Thanks for the clarification.

I'm still fuzzy about the "captures the non-determinism **of a real system**" part though.

Reply:
> "captures the non-determinism of a real system"

what properties of the real system are you interested in modeling? a real world system will have a timer, but timers aren't exact. a real world system may have a target object size, but it is inherently fuzzy because we can't split batches. that's what i mean when i talk about the non-determinism of a real system as it relates to the composition of L0d objects.

Reply:
This is very clear now. May I suggest the following addition?

```suggestion
      // batch size non-determinism of a real system either having reached a maximum
```

Reply:
Committed. thx.